extension/src/functions.rs

```diff
@@ -167,3 +173,223 @@ impl<T: Into<Vec<u8>>> Converter for DefaultConverter<T> {
+	///
+	/// # Parameters
+	/// - `error` - The error to be converted.
+	/// - `env` - The current execution environment.
+	fn convert(error: DispatchError, env: &impl Environment) -> Result<RetVal>;
+}
+
+impl ErrorConverter for () {
+	const LOG_TARGET: &'static str = "pop-chain-extension::converters::error";
+
+	fn convert(error: DispatchError, _env: &impl Environment) -> Result<RetVal> {
+		Err(error)
+	}
+}
+
+// Support tuples of at least one function (required for type resolution) and a maximum of ten.
+#[impl_trait_for_tuples::impl_for_tuples(1, 10)]
+#[tuple_types_custom_trait_bound(Function + Matches)]
+impl<Runtime: pallet_contracts::Config> Function for Tuple {
+	for_tuples!( where #( Tuple: Function<Config=Runtime> )* );
+	type Config = Runtime;
+	type Error = ();
+
+	fn execute(
+		env: &mut (impl Environment<Config = Self::Config> + BufIn + BufOut),
+	) -> Result<RetVal> {
+		// Attempts to match a specified extension/function identifier to its corresponding function, as configured by the runtime.
+		for_tuples!( #(
+            if Tuple::matches(env) {
+                return Tuple::execute(env)
+            }
+        )* );
+
+		// Otherwise returns error indicating an unmatched request.
+		log::error!("no function could be matched");
+		Err(pallet_contracts::Error::<Self::Config>::DecodingFailed.into())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::tests::read_from_buffer;
+	use crate::{
+		mock::INVALID_FUNC_ID,
+		tests::{function_dispatch_call_weight, function_read_state_weight},
+	};
+	use codec::Encode;
+	use frame_support::traits::{Everything, Nothing};
+	use frame_system::Call;
+	use mock::{
+		new_test_ext, DispatchCallEverthingFuncId, Functions, MockEnvironment, MockExt,
+		ReadStateEverthingFuncId, RuntimeCall, RuntimeRead, RuntimeResult, Test,
+	};
+
+	enum AtLeastOneByte {}
+	impl Contains<Vec<u8>> for AtLeastOneByte {
+		fn contains(input: &Vec<u8>) -> bool {
+			input.len() > 0
+		}
+	}
+
+	enum LargerThan100 {}
+	impl Contains<u8> for LargerThan100 {
+		fn contains(input: &u8) -> bool {
+			*input > 100
+		}
+	}
+
+	enum MustBeEven {}
+	impl Contains<u8> for MustBeEven {
+		fn contains(input: &u8) -> bool {
+			*input % 2 == 0
+		}
+	}
+
+	#[test]
+	fn filterting_works() {
+		fn contains<C: Contains<T>, T>(input: T, expected: bool) {
+			assert!(C::contains(&input) == expected);
+		}
+		contains::<Everything, u32>(42, true);
+		contains::<Everything, Vec<u8>>(vec![1, 2, 3, 4], true);
+		contains::<Nothing, u32>(42, false);
+		contains::<Nothing, Vec<u8>>(vec![1, 2, 3, 4], false);
+		contains::<AtLeastOneByte, Vec<u8>>(vec![], false);
+		contains::<AtLeastOneByte, Vec<u8>>(vec![1], true);
+		contains::<AtLeastOneByte, Vec<u8>>(vec![1, 2, 3, 4], true);
+		contains::<LargerThan100, u8>(100, false);
+		contains::<LargerThan100, u8>(101, true);
+		contains::<MustBeEven, u8>(100, true);
+		contains::<MustBeEven, u8>(101, false);
+	}
+
+	#[test]
+	fn execute_dispatch_call_function_works() {
```

Comment:
Whilst you have a unit test for checking that the filter works, there are no tests to ensure that filtering within the functions works. 

Each behaviour of the function should have a different test. For example, for the dispatch function
- decodes
- charges weight
- filters
- dispatches
- adjusts weight
- converts error

Same should apply for the other function.

Reply:
@evilrobot-01  Before we actually have the tests to cover the filtering in the function: https://github.com/r0gue-io/pop-node/pull/240/commits/fcf440b23710b099ec804ef9477bf927f059a15f

But it seems like not very necessary. And the function filtering is does covered in the contract testing. So I am not sure if we want to include the removed tests.